pilot/pkg/xds/ads.go

```diff
@@ -257,6 +265,9 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream discovery.AggregatedD
 			}
 
 		case pushEv := <-con.pushChannel:
+			// Wait for signal from receive thread
+			con.semaphore <- struct{}{}
```

Comment:
If the send failed below, then no ack, so the channel is full and blocking

Reply:
thanks - I had that in my original version - then after cleaning up all the debug, forgot to add that part back in. Thanks !

Reply:
well I decided what I had was not ideal (I was sending a signal in the error path - which could still lead to deadlock). 

Instead, I implemented another semaphore to signal the receive thread to exit. This requires changing the function declaration for `shouldRespond()`.